File: Dante2/Module2.04/Zadanie4.06/array.c
#include <stdio.h>
#include <stdlib.h>
#include "array.h"

int array_create(struct array_t *a, int N)
{
    if(a == NULL || N <= 0)
    {
        return 1;
    }

    a->ptr = malloc(sizeof(int) * N);
    if(!a->ptr)
    {
        return 2;
    }
    a->capacity = N;
    a->size = 0;

    return 0;
}

int array_push_back(struct array_t *a, int value)
{
    if(a == NULL || a->ptr == NULL || a->size > a->capacity ||
       a->capacity <= 0 || a->size < 0)
    {
        return 1;
    }

    if(a->size == a->capacity)
    {
        return 2;
    }

    *(a->ptr + a->size) = value;
    a->size++;

    return 0;
}

void array_display(const struct array_t *a)
{
    if(a == NULL || a->ptr == NULL || a->size > a->capacity ||
       a->capacity <= 0 || a->size <= 0)
    {
        return;
    }

    for(int i = 0; i < a->size; i++)
    {
        printf("%i ",*(a->ptr + i));
    }
}

void array_destroy(struct array_t *a)
{
    if(a == NULL)
    {
        return;
    }

    free(a->ptr);
    a->ptr = NULL;
    a->size = 0;
    a->capacity = 0;
}

File: Dante2/Module2.04/Zadanie4.06/array.rs
#[derive(Default)]
pub struct ArrayT {
    pub array: Vec<i32>,
    pub size: usize,
    pub capacity: usize,
}

pub fn array_create(a: &mut ArrayT, size: usize) {
    a.array = vec![0; size];
    a.size = 0;
    a.capacity = size;
}

pub fn array_push_back(a: &mut ArrayT, value: i32) -> i32 {
    if a.size > a.capacity || a.capacity <= 0 {
        return 1;
    }

    if a.size == a.capacity {
        return 2;
    }

    a.array[a.size] = value;
    a.size += 1;

    0
}

pub fn array_display(a: & ArrayT) {
    if a.size > a.capacity || a.capacity <= 0 {
        return;
    }
    for i in 0..a.size {
        print!("{} ", a.array[i]);
    }
}
